File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cs
public class Solution {
    public IList<string> LetterCombinations(string digits) {    
        List<string> combos = new List<string>();
        
         if (digits.Length == 0) return combos;
        combos.Add("");
        
   string[] digitMap  = {"0", "1" ,"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        for(int i = 0 ; i< digits.Length ; i++ ){
            
                 List<string> temp = new List<string>();
            
            for(int j = 0; j< digitMap[digits[i] - '0'].Length;j++){
                for(int k = 0; k<combos.Count;k++){
                  temp.Add(combos[k] + digitMap[digits[i] - '0'][j]);
                }
            }
            
            combos = temp;
        }
        return combos;
            
}
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:

        """
            first add all characters of first digit
            pop that and add it temp
            for all in the next digit, add that 

        """
        if len(digits) == 0: return []

        d_map = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7":"pqrs",
            "8": "tuv",
            "9": "wxyz",
            "0": " "
        }

        res = []

        def dfs(i, curr):
            if i == len(digits):
                res.append(curr)
            else:
                for ch in d_map[digits[i]]:
                    curr += ch
                    dfs(i + 1, curr)
                    curr = curr[:-1]
        dfs(0, "")
        return res
